### PR TITLE
fix: let vmagent-central chart create file-sd ConfigMap from values

### PR DESCRIPTION
**Pull request by YoonsungNam**
## Summary
- Add a ConfigMap template to the vmagent-central chart that renders file-sd target data from `fileSD.data` values
- Auto-inject target JSON files from `environments/<env>/targets/` via helmfile `exec` — no manual steps
- Remove `fileSD.configMapName` — the chart now owns the ConfigMap lifecycle

## Root cause
The chart Deployment referenced a ConfigMap `vmagent-file-sd` (via `fileSD.configMapName`), but nothing in the deploy pipeline ever created it. Result: `FailedMount` on every corp vmagent-central pod.

## How it works now
```
environments/corp/targets/*.json   (operators edit these)
         │
         ▼
helmfile.yaml.gotmpl  ──(exec cat)──►  fileSD.data values
         │
         ▼
chart configmap-filesd.yaml  ──►  ConfigMap (Helm-managed)
         │
         ▼
deployment.yaml  ──►  volume mount at /etc/vmagent/sd/
```

Operator workflow: edit target JSON files in `targets/`, run `helmfile sync`. One directory, one command.

## Test plan
- [x] `helmfile -e corp template -l name=vmagent-central` — ConfigMap contains all 3 target files
- [x] `helmfile -e homelab template -l name=vmagent-central` — no file-sd resources (no targets/ dir)
- [ ] Deploy to corp cluster and verify vmagent pod starts without FailedMount

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by YoonsungNam** [comment]
Review finding (P2):

This change can still render a broken vmagent deployment when `environments/<env>/targets/*.json` is empty or missing.

Why it matters:
- `helmfile.yaml.gotmpl` only injects `fileSD.data` when JSON target files exist
- `charts/vmagent-central/templates/configmap-filesd.yaml` only renders the `*-file-sd` ConfigMap when `fileSD.data` is non-empty
- `charts/vmagent-central/templates/deployment.yaml` still mounts the `file-sd` volume whenever `fileSD.enabled` is true

That leaves a runtime shape where the Deployment mounts `vmagent-central-vmagent-central-file-sd` but no matching ConfigMap is created, which will fail with `FailedMount`.

Validation:
- `gh pr diff 60`
- `git diff --check origin/dev...origin/fix/vmagent-filesd-configmap`
- throwaway render with empty `targets/`: `helmfile -e corp -l name=vmagent-central template --include-transitive-needs --skip-deps > /tmp/pr60-emptytargets-render.yaml`
- `rg -n "name: vmagent-central-vmagent-central-file-sd|name: file-sd|configMap:" /tmp/pr60-emptytargets-render.yaml -n -C 2`
- `rg -n "kind: ConfigMap|name: vmagent-central-vmagent-central-file-sd" /tmp/pr60-emptytargets-render.yaml -n -C 2`

Suggested fix:
- either fail closed when `fileSD.enabled: true` but no target JSON files are present, or
- gate the volume mount on the same non-empty `fileSD.data` condition used for ConfigMap creation.

**Comment by YoonsungNam** [comment]
## Review

The core approach (chart-managed ConfigMap + helmfile exec injection) is a clean fix for the root cause. Three findings:

### 1. [High] Guard condition mismatch — ConfigMap vs Deployment

The ConfigMap requires `fileSD.enabled AND fileSD.data`, but the Deployment volume mount only checks `fileSD.enabled`. If `enabled: true` but no target files exist yet, the pod gets `FailedMount` — the same failure this PR is fixing.

### 2. [Low] `quote` produces escaped single-line strings in ConfigMap

`{{ $val | quote }}` works but renders multi-line JSON as `"[{\"targets\":[...]}]\n"`. Using `toYaml | nindent` would preserve literal block scalars and produce clean `kubectl get cm -o yaml` output.

### 3. [Nit] Unquoted filename in helmfile exec

The `cat` command interpolates the filename unquoted into a shell string. Filenames with spaces would break. Low risk but easy to harden.

Pushing fixes for all three.

File: charts/vmagent-central/templates/configmap-filesd.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.fileSD.enabled }}
```
Comment (YoonsungNam) [inline charts/vmagent-central/templates/configmap-filesd.yaml]: **[Bug, low]** This renders an empty ConfigMap when `fileSD.enabled: true` but `fileSD.data` is empty (e.g. no `targets/` dir, or operator enables file-sd without providing data). vmagent starts without errors but with zero scrape targets — a silent misconfiguration.

Suggestion: guard on `and .Values.fileSD.enabled .Values.fileSD.data` so a missing targets dir surfaces as a clear `FailedMount` rather than a silent no-op.

File: helmfile.yaml.gotmpl [hunk 2]
```diff
@@ -55,6 +55,17 @@ releases:
...
     values:
       - environments/{{ .Environment.Name }}/vmagent.yaml
+    {{- $targetsDir := printf "environments/%s/targets" .Environment.Name }}
+    {{- if eq (exec "sh" (list "-c" (printf "test -d %s && echo true || echo false" $targetsDir)) | trim) "true" }}
+      - fileSD:
+          data:
+      {{- range (exec "sh" (list "-c" (printf "ls %s/*.json 2>/dev/null | while read f; do basename \"$f\"; done" $targetsDir)) | trim | splitList "\n") }}
+      {{- if ne . "" }}
+            {{ . }}: |
+{{ exec "sh" (list "-c" (printf "cat '%s/%s'" $targetsDir .)) | indent 14 }}
```
Comment (YoonsungNam) [inline helmfile.yaml.gotmpl]: **[Info]** The `cat '%s/%s'` pattern would break if a JSON filename contained a single quote. Low risk since filenames are operator-controlled, but double-quoting would be more robust. Not blocking.

**Review by YoonsungNam — COMMENTED** [review]
Overall the PR is correct and solves the FailedMount issue. Two findings below — one actionable fix, one informational.

**Comment by YoonsungNam** [comment]
## Re-review (post-update)

The guard mismatch (High finding) is resolved. All three guards — ConfigMap template, Deployment volume mount, and Deployment volume — are now consistently `if and .Values.fileSD.enabled .Values.fileSD.data`. Clean approach: don't mount what doesn't exist.

`toYaml` fix retained, filename quoting updated to double quotes. No remaining issues.

LGTM — merging.

**Comment by YoonsungNam** [comment]
**Test fix** (`9611fa0`): The `helm-unittest` failure was due to the test still using the removed `configMapName` field.

Updated in `deployment_test.yaml`:
- Existing test now sets `fileSD.data` and expects the generated name (`RELEASE-NAME-vmagent-central-file-sd`)
- Added volumeMount assertion to the same test
- Added new edge case test: `enabled: true` with empty `data: {}` verifies no volume is mounted

All 25 helm unit tests pass locally.
